File: services/user-api/src/chat/dto/chat.dto.ts
import { ApiProperty, ApiPropertyOptional } from '@nestjs/swagger';
import { IsString, IsArray, IsEnum, IsOptional, IsNumber, IsBoolean } from 'class-validator';

export enum ChatRoomType {
  DIRECT = 'DIRECT',
  GROUP = 'GROUP',
  BOOKING = 'BOOKING',
}

export enum MessageType {
  TEXT = 'TEXT',
  IMAGE = 'IMAGE',
  SYSTEM = 'SYSTEM',
  BOOKING_INVITE = 'BOOKING_INVITE',
  AI_USER = 'AI_USER',
  AI_ASSISTANT = 'AI_ASSISTANT',
}

export class CreateChatRoomDto {
  @ApiProperty({ description: '채팅방 이름' })
  @IsString()
  name: string;

  @ApiProperty({ enum: ChatRoomType, description: '채팅방 타입' })
  @IsEnum(ChatRoomType)
  type: ChatRoomType;

  @ApiProperty({ type: [String], description: '참여자 ID 목록' })
  @IsArray()
  @IsString({ each: true })
  participant_ids: string[];
}

export class SendMessageDto {
  @ApiProperty({ description: '메시지 내용' })
  @IsString()
  content: string;

  @ApiPropertyOptional({ enum: MessageType, default: MessageType.TEXT })
  @IsOptional()
  @IsEnum(MessageType)
  message_type?: MessageType = MessageType.TEXT;
}

export class GetMessagesQueryDto {
  @ApiPropertyOptional({ description: '커서 (이전 응답의 nextCursor)' })
  @IsOptional()
  @IsString()
  cursor?: string;

  @ApiPropertyOptional({ default: 50, description: '가져올 메시지 수' })
  @IsOptional()
  @IsNumber()
  limit?: number = 50;
}

export class GetRoomsQueryDto {
  @ApiPropertyOptional({ default: 1 })
  @IsOptional()
  @IsNumber()
  page?: number = 1;

  @ApiPropertyOptional({ default: 20 })
  @IsOptional()
  @IsNumber()
  limit?: number = 20;
}

export class AddMembersDto {
  @ApiProperty({ type: [String], description: '초대할 사용자 ID 목록' })
  @IsArray()
  @IsString({ each: true })
  user_ids: string[];
}

export class AiChatRequestDto {
  @ApiProperty({ description: 'AI에게 보낼 메시지' })
  @IsString()
  message: string;

  @ApiPropertyOptional({ description: '기존 대화 ID (없으면 새 대화)' })
  @IsOptional()
  @IsString()
  conversationId?: string;

  @ApiPropertyOptional({ description: '사용자 현재 위치 위도' })
  @IsOptional()
  @IsNumber()
  latitude?: number;

  @ApiPropertyOptional({ description: '사용자 현재 위치 경도' })
  @IsOptional()
  @IsNumber()
  longitude?: number;

  // ── 구조화된 선택 필드 (Direct Handling) ──

  @ApiPropertyOptional({ description: '선택한 골프장 ID' })
  @IsOptional()
  @IsString()
  selectedClubId?: string;

  @ApiPropertyOptional({ description: '선택한 골프장 이름' })
  @IsOptional()
  @IsString()
  selectedClubName?: string;

  @ApiPropertyOptional({ description: '선택한 슬롯 ID' })
  @IsOptional()
  @IsString()
  selectedSlotId?: string;

  @ApiPropertyOptional({ description: '선택한 슬롯 시간' })
  @IsOptional()
  @IsString()
  selectedSlotTime?: string;

  @ApiPropertyOptional({ description: '선택한 슬롯 가격' })
  @IsOptional()
  @IsNumber()
  selectedSlotPrice?: number;

  @ApiPropertyOptional({ description: '예약 확인 버튼 클릭' })
  @IsOptional()
  @IsBoolean()
  confirmBooking?: boolean;

  @ApiPropertyOptional({ description: '취소 버튼 클릭' })
  @IsOptional()
  @IsBoolean()
  cancelBooking?: boolean;

  // ── 결제 관련 필드 ──

  @ApiPropertyOptional({ description: '결제방법 ("onsite" | "card")' })
  @IsOptional()
  @IsString()
  paymentMethod?: string;

  @ApiPropertyOptional({ description: '결제 완료 콜백 (Toss 위젯 후)' })
  @IsOptional()
  @IsBoolean()
  paymentComplete?: boolean;

  @ApiPropertyOptional({ description: '결제 성공 여부' })
  @IsOptional()
  @IsBoolean()
  paymentSuccess?: boolean;

  // ── 그룹 예약 필드 ──

  @ApiPropertyOptional({ description: '복수 슬롯 선택 (멀티팀)' })
  @IsOptional()
  @IsArray()
  selectedSlots?: Array<{
    slotId: string;
    slotTime: string;
    courseName: string;
    price: number;
  }>;

  @ApiPropertyOptional({ description: '팀 편성 데이터' })
  @IsOptional()
  teams?: Array<{
    teamNumber: number;
    slotId: string;
    members: Array<{
      userId: number;
      userName: string;
      userEmail: string;
    }>;
  }>;

  @ApiPropertyOptional({ description: '그룹 예약 확인' })
  @IsOptional()
  @IsBoolean()
  confirmGroupBooking?: boolean;

  // ── 분할결제 완료 필드 ──

  @ApiPropertyOptional({ description: '분할결제 완료 콜백' })
  @IsOptional()
  @IsBoolean()
  splitPaymentComplete?: boolean;

  @ApiPropertyOptional({ description: '분할결제 orderId' })
  @IsOptional()
  @IsString()
  splitOrderId?: string;
}

File: apps/user-app-android/app/src/main/java/com/parkgolf/app/data/remote/dto/chat/ChatDto.kt
package com.parkgolf.app.data.remote.dto.chat

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

@Serializable
data class ChatRoomDto(
    val id: String,
    val name: String? = null,
    val type: String, // DIRECT, GROUP, BOOKING
    // API returns 'members', but we also support 'participants' for compatibility
    val members: List<ChatRoomMemberDto>? = null,
    val participants: List<ChatParticipantDto>? = null,
    val lastMessage: ChatMessageDto? = null,
    val unreadCount: Int = 0,
    val createdAt: String,
    val updatedAt: String
)

// API returns 'members' with this structure
@Serializable
data class ChatRoomMemberDto(
    val id: String,
    val roomId: String? = null,
    val userId: Int,
    val userName: String,
    val userEmail: String? = null,
    val joinedAt: String,
    val leftAt: String? = null,
    val isAdmin: Boolean = false,
    val lastReadMessageId: String? = null,
    val lastReadAt: String? = null
)

@Serializable
data class ChatParticipantDto(
    val id: String,
    @SerialName("userId")
    val oduserId: String? = null,
    val userName: String,
    val userEmail: String? = null,
    val profileImageUrl: String? = null,
    val joinedAt: String
)

@Serializable
data class ChatMessageDto(
    val id: String,
    val roomId: String,
    val senderId: Int, // API returns Int
    val senderName: String,
    val content: String,
    // API may return 'type' or 'messageType'
    val messageType: String? = null,
    val type: String? = null, // Fallback field for API compatibility
    val metadata: String? = null,
    val createdAt: String,
    val readBy: List<String>? = null
) {
    // Get the actual message type from either field
    fun getMessageTypeValue(): String = messageType ?: type ?: "TEXT"
}

@Serializable
data class CreateChatRoomRequest(
    val name: String,
    val type: String,
    @SerialName("participant_ids")
    val participantIds: List<String>
)

@Serializable
data class SendMessageRequest(
    val content: String,
    @SerialName("message_type")
    val messageType: String = "TEXT"
)

@Serializable
data class InviteMembersRequest(
    @SerialName("user_ids")
    val userIds: List<String>
)

// API response wrapper for messages endpoint
@Serializable
data class MessagesApiResponse(
    val success: Boolean,
    val data: MessagesData? = null
)

@Serializable
data class MessagesData(
    val messages: List<ChatMessageDto>,
    val hasMore: Boolean = false,
    val nextCursor: String? = null
)

enum class ChatRoomType(val value: String) {
    DIRECT("DIRECT"),
    GROUP("GROUP"),
    BOOKING("BOOKING")
}

enum class MessageType(val value: String) {
    TEXT("TEXT"),
    IMAGE("IMAGE"),
    SYSTEM("SYSTEM"),
    BOOKING_INVITE("BOOKING_INVITE"),
    AI_USER("AI_USER"),
    AI_ASSISTANT("AI_ASSISTANT")
}

// AI Chat DTOs

@Serializable
data class AiChatRequest(
    val message: String,
    val conversationId: String? = null,
    val latitude: Double? = null,
    val longitude: Double? = null,
    val selectedClubId: String? = null,
    val selectedClubName: String? = null,
    val selectedSlotId: String? = null,
    val selectedSlotTime: String? = null,
    val selectedSlotPrice: Int? = null,
    val selectedCourseName: String? = null,
    val confirmBooking: Boolean? = null,
    val cancelBooking: Boolean? = null,
    val paymentMethod: String? = null,
    val paymentComplete: Boolean? = null,
    val paymentSuccess: Boolean? = null,
    // 그룹 예약
    val selectedSlots: List<SelectedSlotDto>? = null,
    val teams: List<TeamDto>? = null,
    val confirmGroupBooking: Boolean? = null,
    // 분할결제 완료
    val splitPaymentComplete: Boolean? = null,
    val splitOrderId: String? = null,
    // 그룹 예약 후속 액션
    val chatRoomId: String? = null,
    val teamMembers: List<TeamMemberDto>? = null,
    val nextTeam: Boolean? = null,
    val finishGroup: Boolean? = null,
    val sendReminder: Boolean? = null
)

@Serializable
data class SelectedSlotDto(
    val slotId: String,
    val slotTime: String,
    val courseName: String,
    val price: Int
)

@Serializable
data class TeamDto(
    val teamNumber: Int,
    val slotId: String,
    val members: List<TeamMemberDto>
)

@Serializable
data class TeamMemberDto(
    val userId: Int,
    val userName: String,
    val userEmail: String
)

@Serializable
data class AiChatActionDto(
    val type: String,
    val data: kotlinx.serialization.json.JsonObject? = null
)

@Serializable
data class AiChatResponseDto(
    val conversationId: String,
    val message: String,
    val state: String,
    val actions: List<AiChatActionDto> = emptyList()
)
